### PR TITLE
Add ability to change locale manually

### PR DESCRIPTION
**Pull request by mrkpatchaa**
Inspired from https://marketplace.visualstudio.com/items?itemName=esskar.vscode-flutter-i18n-json.
Related to #65

### CODE REVIEW

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt [hunk 1]
```diff
@@ -341,11 +337,23 @@ import 'package:flutter/material.dart';
...
+class S implements WidgetsLocalizations {
   const S();
 
   static S current;
 
+  static Locale _locale;
+  static bool _shouldReload = false;
+
+  static set locale(Locale _newLocale) {
+    _shouldReload = true;
```
Comment (noordawod) [inline src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt]: This should be a clever setter: check that the new Locale is different than the active one.

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt [hunk 2]
```diff
@@ -341,11 +337,23 @@ import 'package:flutter/material.dart';
...
-            """class S implements WidgetsLocalizations {
+            """typedef void LocaleChangeCallback(Locale locale);
+
+class S implements WidgetsLocalizations {
   const S();
 
   static S current;
 
+  static Locale _locale;
+  static bool _shouldReload = false;
```
Comment (noordawod) [inline src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt:348]: This variable is redundant, you can detect a new Locale in `bool shouldReload(GeneratedLocalizationsDelegate)` and return the result there.

**Review by noordawod — CHANGES_REQUESTED** [review]
Thanks for the pr @rmkpatchaa.
I left you few comments on the code, please address them first to continue to the next step. Thanks!

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt (hunk 1 above)
Comment (mrkpatchaa) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt]: OK with the clever refresh.
We should have something like that...
```
static set locale(Locale _newLocale) {
    if (getLang(S._locale) != getLang(_newLocale)) {
      _shouldReload = true;
      S._locale = _newLocale;
    }
  }
```

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt (hunk 2 above)
Comment (mrkpatchaa) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt:348]: I don't see how I could detect the new locale in the method.
Unless I add new attributes to the class `GeneratedLocalizationsDelegate` to hold current locale.
The `S` class holds the logic of changing the locale.

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt (hunk 1 above)
Comment (noordawod) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt]: That only checks the language, which is insufficient. The plugin supports also either country-based languages, or languages without a country. So we have to be backward-compatible. Check this method: `bool _isSupported(Locale, bool)`.

So maybe it's best to add an optional 2nd variable (bool withCountry)?

Comment (mrkpatchaa) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt]: The getLang method returns the correct result when locale is declared with country.
`getLang(new Locale("fr"))` returns "fr", and `getLang(new Locale("fr", "FR"))` returns "fr_FR", so I think that comparing `getLang(S._locale)` and `getLang(_newLocale)` is OK.

BTW, we could check if the new lang is supported, before deciding to change the locale.
```
  static set locale(Locale _newLocale) {
    if (S.delegate
            ._isSupported(_newLocale, _newLocale.countryCode?.isNotEmpty) &&
        getLang(S._locale) != getLang(_newLocale)) {
      _shouldReload = true;
      S._locale = _newLocale;
    }
  }
```

Comment (noordawod) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt]: Sounds like a good plan, could you commit the fixes and let's see where we stand?

Comment (mrkpatchaa) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt]: 👍 Changes pushed.

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt [hunk 3]
```diff
@@ -341,11 +337,27 @@ import 'package:flutter/material.dart';
...
+    if (S.delegate
+            ._isSupported(_newLocale, _newLocale.countryCode?.isNotEmpty) &&
+        getLang(S._locale) != getLang(_newLocale)) {
+      _shouldReload = true;
+      S._locale = _newLocale;
+    }
+  }
+
+  static LocaleChangeCallback onLocaleChanged;
+
```
Comment (noordawod) [inline src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt:359]: I suggest rewriting like so:

```dart
  static set locale(Locale _newLocale) {
    _shouldReload = S.delegate._isSupported(_newLocale, _newLocale.countryCode?.isNotEmpty) &&
      getLang(S._locale) != getLang(_newLocale);
    if (_shouldReload) {
      S._locale = _newLocale;
    }
  }
```

**Review by noordawod — CHANGES_REQUESTED** [review]
We're getting there! Left you a single comment about the 'locale()' setter, what do you think?

File: src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt (hunk 3 above)
Comment (mrkpatchaa) [reply src/main/kotlin/eu/long1/flutter/i18n/workers/I18nFileGenerator.kt:359]: Absolutely. It is more clean. I'll make the commit.

**Comment by mrkpatchaa** [comment]
I've updated the code.

**Comment by mrkpatchaa** [comment]
Hi @noordawod any update on this ? 
Thanks.

**Comment by noordawod** [comment]
What do you think @long1eu ?

**Comment by long1eu** [comment]
The way this changes the language is misleading. Normally we rely on the localisation Widget to tell the whole Widget tree what the locale is. This implementation seems more like a hack and doesn't follow the way Flutter normally works.

**Comment by noordawod** [comment]
@long1eu 

I'm in agreement. I was hesitant to merge this in so I'm happy you have doubts too.

Although @rmkpatchaa worked tirelessly to fine-tune the pr according to my input, I vote against merging it in. You?

**Comment by mrkpatchaa** [comment]
@long1eu thanks for the input. Unfortunately I'm not experienced in flutter dev. So yes maybe the solution is not ideal. And maybe someone could help with that. I think it is a feature that is need with a plugin like this one.

**Comment by noordawod** [comment]
I'm closing this because the maintainers think it's out of scope.

**Comment by mrkpatchaa** [comment]
Oh I'm a little bit confused... 😕 
The i18n.dart file is generated by the extension. 

How could we implement the language switch without editing that file ? 
And if edited, it will be overwritten the next time you add a string resource...

Actually I was testing all the modifications I made to the extension by editing my local i18n.dart file, but I know it will be overwritten next time I add a new translation key...

**Comment by mrkpatchaa** [comment]
The point is not about merging my PR. The point is "how do we do that?" if things do not change.

**Comment by long1eu** [comment]
You need to change the locale value of your MaterialApp or WidgetApp. Make
your root Widget stateful and use setState to change this value.

**Comment by mrkpatchaa** [comment]
Ok, I've tried with a custom class and it worked.
One thing though
If we want to test on supported locales, the _isSupported method is always called with true.
```
@override
  bool isSupported(Locale locale) => _isSupported(locale, true);
```
This could be written 
```
@override
  bool isSupported(Locale locale) => _isSupported(locale, locale.countryCode?.isNotEmpty);
```
Thanks.

**Comment by noordawod** [comment]
I believe this was done in this fashion because of backward-compatibility. It may break apps if we change the implementation now.

**Comment by mrkpatchaa** [comment]
OK. Thanks for the input.
BTW it's mentioned that you're working on a VSCode extension. Is it ready yet for testing ? 
Thanks.

**Comment by long1eu** [comment]
Not yet unfortunately. Do you have knowledge of building a vccode extension?

**Comment by mrkpatchaa** [comment]
Unfortunately no. 
But maybe I could make some time to learn by helping on this.

**Pull request closed without merging**